src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -236,103 +246,75 @@ void executeQueryWithParallelReplicas(
-                "Limit for bytes to send or receive over network exceeded.",
-                user_level_throttler);
-    }
-    else
-        throttler = user_level_throttler;
-
-
-    std::vector<QueryPlanPtr> plans;
-    SelectStreamFactory::Shards remote_shards;
-    size_t shards = query_info.getCluster()->getShardCount();
-
-    for (const auto & shard_info : query_info.getCluster()->getShardsInfo())
-    {
-        ASTPtr query_ast_for_shard;
-        if (query_info.optimized_cluster && settings.optimize_skip_unused_shards_rewrite_in && shards > 1)
-        {
-            query_ast_for_shard = query_ast->clone();
-
-            OptimizeShardingKeyRewriteInVisitor::Data visitor_data{
-                sharding_key_expr,
-                sharding_key_expr->getSampleBlock().getByPosition(0).type,
-                sharding_key_column_name,
-                shard_info,
-                not_optimized_cluster->getSlotToShard(),
-            };
-            OptimizeShardingKeyRewriteInVisitor visitor(visitor_data);
-            visitor.visit(query_ast_for_shard);
-        }
-        else
-            query_ast_for_shard = query_ast;
-
-        stream_factory.createForShardWithParallelReplicas(
-            shard_info, query_ast_for_shard, main_table, context, static_cast<UInt32>(shards), plans, remote_shards);
-    }
-
-    Scalars scalars = context->hasQueryContext() ? context->getQueryContext()->getScalars() : Scalars{};
-    scalars.emplace(
-        "_shard_count", Block{{DataTypeUInt32().createColumnConst(1, shards), std::make_shared<DataTypeUInt32>(), "_shard_count"}});
-    auto external_tables = context->getExternalTables();
-
-    if (!remote_shards.empty())
-    {
-        auto new_context = Context::createCopy(context);
-
-        for (const auto & shard : remote_shards)
-        {
-            auto read_from_remote = std::make_unique<ReadFromParallelRemoteReplicasStep>(
-                shard.coordinator,
-                shard,
-                shard.header,
-                processed_stage,
-                main_table,
-                table_func_ptr,
-                new_context,
-                throttler,
-                scalars,
-                external_tables,
-                &Poco::Logger::get("ReadFromParallelRemoteReplicasStep"),
-                query_info.storage_limits);
-
-            auto remote_plan = std::make_unique<QueryPlan>();
-            remote_plan->addStep(std::move(read_from_remote));
-            remote_plan->addInterpreterContext(new_context);
-            plans.emplace_back(std::move(remote_plan));
-        }
-    }
-
-    if (plans.empty())
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "No plans were generated for reading from Distributed. This is a bug");
-
-    if (plans.size() == 1)
+    if (not_optimized_cluster->getShardsInfo().size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cluster for parallel replicas should consist only from one shard");
```

Comment:
Probably not a LOGICAL_ERROR, but user's error.
Please, add a number of shards to the exception message.